Zephyr example does not build successfully (ESF-182)

### Port

Zephyr

### Target chip

ESP32

### Hardware Configuration

esp32_devkitc_wroom

### Log output

```plain
root@52b956611206:/workdir# west build -p -b esp32_devkitc_wroom esp-serial-flasher/examples/zephyr_example/
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /workdir/esp-serial-flasher/examples/zephyr_example
-- CMake version: 3.25.1
-- Found Python3: /opt/venv/bin/python3 (found suitable version "3.11.2", minimum required is "3.10") found components: Interpreter 
-- Cache files will be written to: /root/.cache/zephyr
-- Zephyr version: 4.0.99 (/workdir/zephyr)
-- Found west (found suitable version "1.3.0", minimum required is "0.14.0")
CMake Warning at /workdir/zephyr/cmake/modules/boards.cmake:110 (message):
  Deprecated BOARD=esp32_devkitc_wroom specified, board automatically changed
  to: esp32_devkitc_wroom/esp32/procpu.
Call Stack (most recent call first):
  /workdir/zephyr/cmake/modules/zephyr_default.cmake:133 (include)
  /workdir/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  /workdir/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  CMakeLists.txt:5 (find_package)


-- Board: esp32_devkitc_wroom, qualifiers: esp32/procpu
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.17.0 (/root/zephyr-sdk-0.17.0)
-- Found toolchain: zephyr 0.17.0 (/root/zephyr-sdk-0.17.0)
-- Found Dtc: /usr/bin/dtc (found suitable version "1.6.1", minimum required is "1.4.6") 
-- Found BOARD.dts: /workdir/zephyr/boards/espressif/esp32_devkitc_wroom/esp32_devkitc_wroom_procpu.dts
-- Found devicetree overlay: /workdir/esp-serial-flasher/examples/zephyr_example/socs/esp32_procpu.overlay
-- Generated zephyr.dts: /workdir/build/zephyr/zephyr.dts
-- Generated pickled edt: /workdir/build/zephyr/edt.pickle
-- Generated zephyr.dts: /workdir/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: /workdir/build/zephyr/include/generated/zephyr/devicetree_generated.h
-- Including generated dts.cmake file: /workdir/build/zephyr/dts.cmake
CMake Warning at /workdir/zephyr/cmake/modules/dts.cmake:426 (message):
  dtc raised one or more warnings:

  /workdir/build/zephyr/zephyr.dts:46.39-53.5: Warning (interrupt_provider):
  /soc/interrupt-controller@3ff00104: Missing #address-cells in interrupt
  provider

Call Stack (most recent call first):
  /workdir/zephyr/cmake/modules/zephyr_default.cmake:133 (include)
  /workdir/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  /workdir/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  CMakeLists.txt:5 (find_package)



warning: The choice symbol CONSOLE_GETCHAR (defined at subsys/console/Kconfig:16) was selected (set
=y), but no symbol ended up as the choice selection. See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_CONSOLE_GETCHAR and/or look up
CONSOLE_GETCHAR in the menuconfig/guiconfig interface. The Application Development Primer, Setting
Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
too.

Parsing /workdir/zephyr/Kconfig
Loaded configuration '/workdir/zephyr/boards/espressif/esp32_devkitc_wroom/esp32_devkitc_wroom_procpu_defconfig'
Merged configuration '/workdir/esp-serial-flasher/examples/zephyr_example/prj.conf'
Configuration saved to '/workdir/build/zephyr/.config'
Kconfig header saved to '/workdir/build/zephyr/include/generated/zephyr/autoconf.h'
-- Found GnuLd: /root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd (found version "2.38") 
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/xtensa-espressif_esp32_zephyr-elf-gcc
esptool path: /workdir/modules/hal/espressif/tools/esptool_py/esptool.py
-- Configuring done
-- Generating done
-- Build files have been written to: /workdir/build
-- west build: building application
[3/202] Preparing syscall dependency handling

[4/202] Generating include/generated/zephyr/version.h
-- Zephyr version: 4.0.99 (/workdir/zephyr), build: v4.0.0-609-g28796076d840
[15/202] Building C object CMakeFiles/app.dir/workdir/esp-serial-flasher/examples/common/example_common.c.obj
FAILED: CMakeFiles/app.dir/workdir/esp-serial-flasher/examples/common/example_common.c.obj 
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/xtensa-espressif_esp32_zephyr-elf-gcc -DBOOTLOADER_BUILD -DCONFIG_APP_BUILD_USE_FLASH_SECTIONS -DCONFIG_ESP_SYSTEM_CHECK_INT_LEVEL_5 -DESP_PLATFORM -DKERNEL -DK_HEAP_MEM_POOL_SIZE=4096 -DSERIAL_FLASHER_INTERFACE_UART -D_ANSI_SOURCE -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR__=1 -I/workdir/esp-serial-flasher/examples/zephyr_example/../common -I/workdir/esp-serial-flasher/examples/zephyr_example/../binaries -I/workdir/build/zephyr/include/generated/zephyr -I/workdir/zephyr/include -I/workdir/build/zephyr/include/generated -I/workdir/zephyr/soc/espressif -I/workdir/zephyr/lib/libc/newlib/include -I/workdir/zephyr/soc/espressif/common/include -I/workdir/zephyr/soc/espressif/esp32/. -I/workdir/modules/hal/espressif/zephyr/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/include/bt -I/workdir/modules/hal/espressif/zephyr/esp32/../esp_shared/include -I/workdir/modules/hal/espressif/zephyr/esp32/../esp_shared/components/include -I/workdir/modules/hal/espressif/zephyr/esp32/../port/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/efuse/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/efuse/private_include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/efuse/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/efuse/esp32/private_include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_common/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/dma -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/include/esp_private -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/include/hal -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/include/soc -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/include/soc/esp32 -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/port/esp32 -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/port/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32 -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/include/esp32 -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_system/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_system/include/esp_private -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_system/port/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_system/port/include/private -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/hal/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/hal/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/hal/platform_port/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/log/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/soc/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/soc/esp32/include/soc -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/soc/esp32/ld -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/soc/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/xtensa/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/xtensa/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/xtensa/include/esp_private -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/xtensa/include/xtensa -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_timer/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_timer/private_include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/driver/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/driver/deprecated -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/driver/gpio/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/driver/uart/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/driver/touch_sensor/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/driver/touch_sensor/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/driver/spi/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/spi_flash/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/spi_flash/include/spi_flash -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_pm/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/bootloader_support/bootloader_flash/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/bootloader_support/private_include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/bootloader_support/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/heap/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_psram/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_mm/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_wifi/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_event/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_event/private_include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_phy/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_phy/include/esp32 -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_phy/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_wifi/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_netif/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/esp_supplicant/src -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/esp_supplicant/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/include/esp_supplicant -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/port/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/src -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/src/crypto -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/src/eap_peer -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/src/utils -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_coex/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/mbedtls/port/include -I/workdir/modules/hal/espressif/zephyr/esp32/../port/include/boot -isystem /workdir/zephyr/lib/libc/common/include -fno-strict-aliasing -Os -imacros /workdir/build/zephyr/include/generated/zephyr/autoconf.h -fno-common -g -gdwarf-4 -fdiagnostics-color=always --sysroot=/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/xtensa-espressif_esp32_zephyr-elf -imacros /workdir/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/workdir/esp-serial-flasher/examples/zephyr_example=CMAKE_SOURCE_DIR -fmacro-prefix-map=/workdir/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/workdir=WEST_TOPDIR -ffunction-sections -fdata-sections -mlongcalls -fstrict-volatile-bitfields -std=c99 -MD -MT CMakeFiles/app.dir/workdir/esp-serial-flasher/examples/common/example_common.c.obj -MF CMakeFiles/app.dir/workdir/esp-serial-flasher/examples/common/example_common.c.obj.d -o CMakeFiles/app.dir/workdir/esp-serial-flasher/examples/common/example_common.c.obj -c /workdir/esp-serial-flasher/examples/common/example_common.c
/workdir/esp-serial-flasher/examples/common/example_common.c:22:10: fatal error: esp_loader_io.h: No such file or directory
   22 | #include "esp_loader_io.h"
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
[17/202] Building C object CMakeFiles/app.dir/src/main.c.obj
FAILED: CMakeFiles/app.dir/src/main.c.obj 
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/xtensa-espressif_esp32_zephyr-elf-gcc -DBOOTLOADER_BUILD -DCONFIG_APP_BUILD_USE_FLASH_SECTIONS -DCONFIG_ESP_SYSTEM_CHECK_INT_LEVEL_5 -DESP_PLATFORM -DKERNEL -DK_HEAP_MEM_POOL_SIZE=4096 -DSERIAL_FLASHER_INTERFACE_UART -D_ANSI_SOURCE -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR__=1 -I/workdir/esp-serial-flasher/examples/zephyr_example/../common -I/workdir/esp-serial-flasher/examples/zephyr_example/../binaries -I/workdir/build/zephyr/include/generated/zephyr -I/workdir/zephyr/include -I/workdir/build/zephyr/include/generated -I/workdir/zephyr/soc/espressif -I/workdir/zephyr/lib/libc/newlib/include -I/workdir/zephyr/soc/espressif/common/include -I/workdir/zephyr/soc/espressif/esp32/. -I/workdir/modules/hal/espressif/zephyr/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/include/bt -I/workdir/modules/hal/espressif/zephyr/esp32/../esp_shared/include -I/workdir/modules/hal/espressif/zephyr/esp32/../esp_shared/components/include -I/workdir/modules/hal/espressif/zephyr/esp32/../port/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/efuse/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/efuse/private_include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/efuse/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/efuse/esp32/private_include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_common/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/dma -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/include/esp_private -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/include/hal -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/include/soc -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/include/soc/esp32 -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/port/esp32 -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/port/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32 -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/include/esp32 -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_system/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_system/include/esp_private -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_system/port/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_system/port/include/private -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/hal/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/hal/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/hal/platform_port/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/log/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/soc/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/soc/esp32/include/soc -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/soc/esp32/ld -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/soc/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/xtensa/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/xtensa/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/xtensa/include/esp_private -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/xtensa/include/xtensa -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_timer/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_timer/private_include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/driver/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/driver/deprecated -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/driver/gpio/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/driver/uart/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/driver/touch_sensor/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/driver/touch_sensor/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/driver/spi/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/spi_flash/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/spi_flash/include/spi_flash -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_pm/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/bootloader_support/bootloader_flash/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/bootloader_support/private_include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/bootloader_support/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/heap/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_psram/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_mm/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_wifi/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_event/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_event/private_include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_phy/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_phy/include/esp32 -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_phy/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_wifi/esp32/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_netif/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/esp_supplicant/src -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/esp_supplicant/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/include/esp_supplicant -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/port/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/src -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/src/crypto -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/src/eap_peer -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/src/utils -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_coex/include -I/workdir/modules/hal/espressif/zephyr/esp32/../../components/mbedtls/port/include -I/workdir/modules/hal/espressif/zephyr/esp32/../port/include/boot -isystem /workdir/zephyr/lib/libc/common/include -fno-strict-aliasing -Os -imacros /workdir/build/zephyr/include/generated/zephyr/autoconf.h -fno-common -g -gdwarf-4 -fdiagnostics-color=always --sysroot=/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/xtensa-espressif_esp32_zephyr-elf -imacros /workdir/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/workdir/esp-serial-flasher/examples/zephyr_example=CMAKE_SOURCE_DIR -fmacro-prefix-map=/workdir/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/workdir=WEST_TOPDIR -ffunction-sections -fdata-sections -mlongcalls -fstrict-volatile-bitfields -std=c99 -MD -MT CMakeFiles/app.dir/src/main.c.obj -MF CMakeFiles/app.dir/src/main.c.obj.d -o CMakeFiles/app.dir/src/main.c.obj -c /workdir/esp-serial-flasher/examples/zephyr_example/src/main.c
/workdir/esp-serial-flasher/examples/zephyr_example/src/main.c:20:10: fatal error: zephyr_port.h: No such file or directory
   20 | #include <zephyr_port.h>
      |          ^~~~~~~~~~~~~~~
compilation terminated.
[24/202] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/cbprintf_complete.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /workdir/build
```


### More Information

I created this Dockerfile to test building in a clean environment. It seems like there's an issue with examples/zephyr_example/CMakeLists.txt but I'm not exactly sure.

```Dockerfile
FROM debian:stable-slim

# OS dependencies and packages

RUN \
  apt-get -y update \
  && apt-get -y install --no-install-recommends \
  ca-certificates \
  cmake \
  device-tree-compiler \
  git \
  ninja-build \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/*

# Python

ENV VIRTUAL_ENV=/opt/venv
RUN \
  apt-get -y update \
  && apt-get -y install --no-install-recommends \
  python3 \
  python3-pip \
  python3-venv \
  && python3 -m venv $VIRTUAL_ENV
ENV PATH="$VIRTUAL_ENV/bin:$PATH"

# West

RUN pip install --no-cache-dir wheel west

WORKDIR /workdir

# Zephyr setup

RUN \
  west init \
  && west update --narrow -o=--depth=1 \
  && west zephyr-export \
  && pip install -r zephyr/scripts/requirements.txt

# Zephyr SDK

RUN \
  apt-get -y update \
  && apt-get -y install --no-install-recommends \
  wget \
  xz-utils \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/* \
  && west sdk install -t xtensa-espressif_esp32_zephyr-elf -H

# ESP Flasher

RUN \
  apt-get -y update \
  && apt-get -y install --no-install-recommends \
  unzip \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/* \
  && git clone https://github.com/espressif/esp-serial-flasher.git \
  && wget https://dl.espressif.com/esp-at/firmwares/esp32/ESP32-PICO-D4/ESP32-PICO-D4-AT-V3.2.1.0.zip \
  && unzip ESP32-PICO-D4-AT-V3.2.1.0.zip \
  && mv ESP32-PICO-D4-AT-V3.2.1.0/ESP32-PICO-D4-AT-V3.2.1.0/esp-at.bin esp-serial-flasher/examples/binaries/ESP32_AT_Firmware/
# && west build -p -b esp32_devkitc_wroom esp-serial-flasher/examples/zephyr_example/
```

Adding the `include` & `port` directories fixed the CMake issues:

```
zephyr_library_include_directories(
  ../common
  ../binaries
  ../../include
  ../../port
  )
```

But I'm getting linking issues:

```
[197/202] Linking C executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /workdir/build/zephyr/zephyr_pre0.map 
: && /root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/xtensa-espressif_esp32_zephyr-elf-gcc  -gdwarf-4 -gdwarf-4 zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/xtensa/core/offsets/offsets.c.obj  -T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/workdir/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/xtensa/core/libarch__xtensa__core.a  zephyr/lib/libc/newlib/liblib__libc__newlib.a  zephyr/lib/libc/common/liblib__libc__common.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L/workdir/build/zephyr  zephyr/arch/common/libisr_tables.a  -fuse-ld=bfd  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn  -Wl,-no-pie  -L"/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/xtensa-espressif_esp32_zephyr-elf"/lib/.  -lgcc  -T/workdir/modules/hal/espressif/zephyr/esp32/src/linker/esp32.rom.alias.ld  -T/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.ld  -T/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.api.ld  -T/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.libgcc.ld  -T/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.newlib-data.ld  -T/workdir/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.newlib-funcs.ld  -T/workdir/modules/hal/espressif/zephyr/esp32/../../components/soc/esp32/ld/esp32.peripherals.ld  -Wl,--wrap=longjmp  -lc -L"/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0" -lm -lc -lgcc -lc && cd /workdir/build/zephyr && /usr/bin/cmake -E true
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: app/libapp.a(example_common.c.obj):(.literal.connect_to_target+0x28): undefined reference to `esp_loader_connect'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: app/libapp.a(example_common.c.obj):(.literal.connect_to_target+0x2c): undefined reference to `esp_loader_get_target'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: app/libapp.a(example_common.c.obj):(.literal.connect_to_target+0x30): undefined reference to `esp_loader_change_transmission_rate'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: app/libapp.a(example_common.c.obj):(.literal.connect_to_target+0x34): undefined reference to `loader_port_change_transmission_rate'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: app/libapp.a(example_common.c.obj):(.literal.flash_binary+0x24): undefined reference to `esp_loader_flash_start'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: app/libapp.a(example_common.c.obj):(.literal.flash_binary+0x28): undefined reference to `esp_loader_flash_write'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: app/libapp.a(example_common.c.obj): in function `connect_to_target':
/workdir/esp-serial-flasher/examples/common/example_common.c:263: undefined reference to `esp_loader_connect'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: /workdir/esp-serial-flasher/examples/common/example_common.c:277: undefined reference to `esp_loader_get_target'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: /workdir/esp-serial-flasher/examples/common/example_common.c:280: undefined reference to `esp_loader_change_transmission_rate'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: /workdir/esp-serial-flasher/examples/common/example_common.c:285: undefined reference to `loader_port_change_transmission_rate'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: app/libapp.a(example_common.c.obj): in function `flash_binary':
/workdir/esp-serial-flasher/examples/common/example_common.c:354: undefined reference to `esp_loader_flash_start'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: /workdir/esp-serial-flasher/examples/common/example_common.c:371: undefined reference to `esp_loader_flash_write'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: app/libapp.a(main.c.obj):(.literal.main+0x24): undefined reference to `loader_port_zephyr_init'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: app/libapp.a(main.c.obj):(.literal.main+0x28): undefined reference to `esp_loader_reset_target'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: app/libapp.a(main.c.obj): in function `main':
/workdir/esp-serial-flasher/examples/zephyr_example/src/main.c:62: undefined reference to `loader_port_zephyr_init'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: /workdir/esp-serial-flasher/examples/zephyr_example/src/main.c:66: undefined reference to `esp_loader_get_target'
/root/zephyr-sdk-0.17.0/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: /workdir/esp-serial-flasher/examples/zephyr_example/src/main.c:73: undefined reference to `esp_loader_reset_target'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /workdir/build
```

--- a/src/esp_loader.c
+++ b/src/esp_loader.c
@@ -609,7 +609,7 @@ static esp_loader_error_t flash_read_stub(uint8_t *dest, uint32_t address, uint3
 esp_loader_error_t esp_loader_flash_read(uint8_t *dest, uint32_t address, uint32_t length)
 {
     RETURN_ON_ERROR(init_flash_params());
-    if (address + length >= s_target_flash_size) {
+    if (address + length > s_target_flash_size) {
         return ESP_LOADER_ERROR_IMAGE_SIZE;
     }
 
@@ -810,7 +810,7 @@ esp_loader_error_t esp_loader_flash_verify_known_md5(uint32_t address,
 
     RETURN_ON_ERROR(init_flash_params());
 
-    if (address + size >= s_target_flash_size) {
+    if (address + size > s_target_flash_size) {
         return ESP_LOADER_ERROR_IMAGE_SIZE;
     }
 

